File: src/cli/commands/create.ts
import { existsSync, mkdirSync, writeFileSync } from "node:fs";
import { dirname } from "node:path";
import {
	FIELDS,
	type Frontmatter,
	isSpec,
	parseFile,
	resolveSpecPath,
	SPEC_STATUSES,
	SPEC_TYPES,
	SpecIndex,
	type SpecType,
	serializeFrontmatter,
} from "../../core/index.ts";
import { parseCommandArgs, rootFrom, UsageError } from "../args.ts";
import { print, printJson } from "../out.ts";

const SCAFFOLD_HEADINGS: Record<SpecType, readonly string[]> = {
	"module-design": ["Responsibility", "Boundary"],
	"submodule-design": ["Responsibility", "Boundary"],
	"architecture-design": ["Drivers", "Decisions", "Invariants", "Out of scope"],
	"goal-and-requirements": ["Goal", "Scope"],
	"task-spec": ["Purpose", "Open items"],
};

function scaffoldBody(type: SpecType): string {
	return SCAFFOLD_HEADINGS[type].map((heading) => `## ${heading}\n`).join("\n");
}

export async function run(argv: string[]): Promise<number> {
	const { values, positionals } = parseCommandArgs(argv, {
		id: { type: "string" },
		type: { type: "string" },
		title: { type: "string" },
		status: { type: "string" },
		parent: { type: "string" },
		"depends-on": { type: "string", multiple: true },
		references: { type: "string", multiple: true },
		implements: { type: "string", multiple: true },
		covers: { type: "string", multiple: true },
		tags: { type: "string", multiple: true },
	});
	const path = positionals[0];
	if (path === undefined) throw new UsageError("create requires a path argument.");
	const id = values.id;
	if (id === undefined) throw new UsageError("--id is required.");
	const type = SPEC_TYPES.find((candidate) => candidate === values.type);
	if (type === undefined) throw new UsageError(`--type must be one of: ${SPEC_TYPES.join(", ")}`);
	const title = values.title;
	if (title === undefined) throw new UsageError("--title is required.");
	const status = SPEC_STATUSES.find((candidate) => candidate === values.status);
	if (values.status !== undefined && status === undefined) {
		throw new UsageError(`--status must be one of: ${SPEC_STATUSES.join(", ")}`);
	}

	const root = rootFrom(values);
	const resolved = resolveSpecPath(root, path);
	if ("error" in resolved) {
		process.stderr.write(`${resolved.error}\n`);
		return 2;
	}
	const { rel, abs } = resolved;
	const index = new SpecIndex(root);
	if (existsSync(abs)) {
		process.stderr.write(`File already exists: ${rel}\n`);
		return 2;
	}
	if (index.graph().nodes.has(id)) {
		process.stderr.write(`Spec id "${id}" is already in use.\n`);
		return 2;
	}

	const frontmatter: Frontmatter = { [FIELDS.id]: id, [FIELDS.type]: type };
	if (status !== undefined) frontmatter[FIELDS.status] = status;
	frontmatter[FIELDS.title] = title;
	if (values.parent !== undefined) frontmatter[FIELDS.parent] = values.parent;
	if (values["depends-on"]?.length) frontmatter[FIELDS.dependsOn] = values["depends-on"];
	if (values.references?.length) frontmatter[FIELDS.references] = values.references;
	if (values.implements?.length) frontmatter[FIELDS.implements] = values.implements;
	if (values.covers?.length) frontmatter[FIELDS.covers] = values.covers;
	if (values.tags?.length) frontmatter[FIELDS.tags] = values.tags;

	const content = `${serializeFrontmatter(frontmatter)}\n${scaffoldBody(type)}`;
	if (!isSpec(parseFile(content).frontmatter)) {
		process.stderr.write(
			`Refusing to write ${rel}: the frontmatter would not be a spec (id and type must be non-empty).\n`,
		);
		return 2;
	}
	try {
		mkdirSync(dirname(abs), { recursive: true });
		writeFileSync(abs, content, { encoding: "utf8", flag: "wx" });
	} catch (err) {
		process.stderr.write(
			`Failed to write ${rel}: ${err instanceof Error ? err.message : String(err)}\n`,
		);
		return 2;
	}
	const details = { path: rel, id };
	if (values.json === true) printJson(details);
	else print(`Created ${rel} (id: ${id}).`);
	return 0;
}

File: src/cli/commands/create.zig
const std = @import("std");
const graph_mod = @import("../../core/graph.zig");
const parse = @import("../../core/parse.zig");
const store = @import("../../core/store.zig");
const args_mod = @import("../args.zig");
const context = @import("../context.zig");
const json = @import("../../json.zig");

const command_specs = [_]args_mod.Spec{
    .{ .name = "id" },
    .{ .name = "type" },
    .{ .name = "title" },
    .{ .name = "status" },
    .{ .name = "parent" },
    .{ .name = "depends-on", .multiple = true },
    .{ .name = "references", .multiple = true },
    .{ .name = "implements", .multiple = true },
    .{ .name = "covers", .multiple = true },
    .{ .name = "tags", .multiple = true },
};

fn scaffoldBody(arena: std.mem.Allocator, spec_type: []const u8) anyerror![]const u8 {
    const headings: []const []const u8 = if (std.mem.eql(u8, spec_type, "module-design") or std.mem.eql(u8, spec_type, "submodule-design"))
        &.{ "Responsibility", "Boundary" }
    else if (std.mem.eql(u8, spec_type, "architecture-design"))
        &.{ "Drivers", "Decisions", "Invariants", "Out of scope" }
    else if (std.mem.eql(u8, spec_type, "goal-and-requirements"))
        &.{ "Goal", "Scope" }
    else
        &.{ "Purpose", "Open items" };
    var out: std.Io.Writer.Allocating = .init(arena);
    for (headings, 0..) |heading, index| {
        if (index > 0) try out.writer.writeByte('\n');
        try out.writer.print("## {s}\n", .{heading});
    }
    return out.written();
}

pub fn run(ctx: *context.Context) anyerror!u8 {
    const parsed = try args_mod.parse(ctx.arena, ctx.argv, &command_specs);
    const path = if (parsed.positionals.items.len > 0) parsed.positionals.items[0] else
        return args_mod.usageError(ctx.arena, "create requires a path argument.", .{});
    const id = parsed.stringOf("id") orelse return args_mod.usageError(ctx.arena, "--id is required.", .{});
    var spec_type: ?[]const u8 = null;
    if (parsed.stringOf("type")) |wanted| {
        for (parse.SPEC_TYPES) |candidate| {
            if (std.mem.eql(u8, candidate, wanted)) spec_type = candidate;
        }
    }
    const resolved_type = spec_type orelse
        return args_mod.usageError(ctx.arena, "--type must be one of: goal-and-requirements, architecture-design, module-design, submodule-design, task-spec", .{});
    const title = parsed.stringOf("title") orelse return args_mod.usageError(ctx.arena, "--title is required.", .{});
    var status: ?[]const u8 = null;
    if (parsed.stringOf("status")) |wanted| {
        for (parse.SPEC_STATUSES) |candidate| {
            if (std.mem.eql(u8, candidate, wanted)) status = candidate;
        }
        if (status == null) {
            return args_mod.usageError(ctx.arena, "--status must be one of: draft, active, stale, done, deprecated", .{});
        }
    }

    const root = try args_mod.rootFrom(ctx.arena, ctx.io, parsed.stringOf("root"));
    const resolution = try store.resolveSpecPath(ctx.arena, ctx.io, root, path);
    const resolved = switch (resolution) {
        .err => |message| {
            try ctx.out.err(message);
            return 2;
        },
        .resolved => |value| value,
    };
    if (std.Io.Dir.cwd().statFile(ctx.io, resolved.abs, .{}) catch null) |_| {
        try ctx.out.err(try std.fmt.allocPrint(ctx.arena, "File already exists: {s}", .{resolved.rel}));
        return 2;
    }
    var index = store.SpecIndex.init(ctx.arena, ctx.io, root);
    if ((try index.graph()).nodes.contains(id)) {
        try ctx.out.err(try std.fmt.allocPrint(ctx.arena, "Spec id \"{s}\" is already in use.", .{id}));
        return 2;
    }

    var entries = std.ArrayListUnmanaged(parse.Entry).empty;
    try entries.append(ctx.arena, .{ .key = parse.ID, .value = .{ .scalar = id } });
    try entries.append(ctx.arena, .{ .key = parse.TYPE, .value = .{ .scalar = resolved_type } });
    if (status) |status_value| {
        try entries.append(ctx.arena, .{ .key = parse.STATUS, .value = .{ .scalar = status_value } });
    }
    try entries.append(ctx.arena, .{ .key = parse.TITLE, .value = .{ .scalar = title } });
    if (parsed.stringOf("parent")) |parent| {
        try entries.append(ctx.arena, .{ .key = parse.PARENT, .value = .{ .scalar = parent } });
    }
    inline for ([_][]const u8{ "depends-on", "references", "implements", "covers", "tags" }) |field| {
        const items = parsed.listOf(field);
        if (items.len > 0) {
            try entries.append(ctx.arena, .{ .key = field, .value = .{ .list = items } });
        }
    }
    const frontmatter = parse.Frontmatter{ .entries = entries.items };

    const content = try std.fmt.allocPrint(ctx.arena, "{s}\n{s}", .{
        try parse.serializeFrontmatter(ctx.arena, frontmatter),
        try scaffoldBody(ctx.arena, resolved_type),
    });
    const written = try parse.parseFile(ctx.arena, content);
    if (!parse.isSpec(written.frontmatter)) {
        try ctx.out.err(try std.fmt.allocPrint(
            ctx.arena,
            "Refusing to write {s}: the frontmatter would not be a spec (id and type must be non-empty).",
            .{resolved.rel},
        ));
        return 2;
    }

    if (std.fs.path.dirname(resolved.abs)) |parent_dir| {
        std.Io.Dir.cwd().createDirPath(ctx.io, parent_dir) catch {};
    }
    var file = std.Io.Dir.cwd().createFile(ctx.io, resolved.abs, .{ .exclusive = true }) catch |err| {
        try ctx.out.err(try std.fmt.allocPrint(ctx.arena, "Failed to write {s}: {s}", .{ resolved.rel, @errorName(err) }));
        return 2;
    };
    defer file.close(ctx.io);
    file.writeStreamingAll(ctx.io, content) catch |err| {
        try ctx.out.err(try std.fmt.allocPrint(ctx.arena, "Failed to write {s}: {s}", .{ resolved.rel, @errorName(err) }));
        return 2;
    };

    if (parsed.boolOf("json")) {
        const fields = try ctx.arena.dupe(json.Field, &[_]json.Field{
            .{ .key = "path", .value = .{ .string = resolved.rel } },
            .{ .key = "id", .value = .{ .string = id } },
        });
        try ctx.out.print(try json.stringify(ctx.arena, .{ .object = fields }));
    } else {
        try ctx.out.print(try std.fmt.allocPrint(ctx.arena, "Created {s} (id: {s}).", .{ resolved.rel, id }));
    }
    return 0;
}
